<h2 align='center'>Dropout Regularization In Deep Neural Network</h2>

Dataset basically about if the sonar signal would bounce off metal cyllinder or roughly cyllindrical rocks.

Dataset: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [116]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from dython.nominal import associations
pd.set_option('display.max_columns', None) ##just to enable pandas to show all the available columns

In [117]:
df = pd.read_csv("./sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
77   0.0336  0.0294  0.0476  0.0539  0.0794  0.0804  0.1136  0.1228  0.1235   
142  0.0526  0.0563  0.1219  0.1206  0.0246  0.1022  0.0539  0.0439  0.2291   
102  0.0587  0.1210  0.1268  0.1498  0.1436  0.0561  0.0832  0.0672  0.1372   
130  0.0443  0.0446  0.0235  0.1008  0.2252  0.2611  0.2061  0.1668  0.1801   
36   0.0094  0.0166  0.0398  0.0359  0.0681  0.0706  0.1020  0.0893  0.0381   

         9       10      11      12      13      14      15      16      17  \
77   0.0842  0.0357  0.0689  0.1705  0.3257  0.4602  0.6225  0.7327  0.7843   
142  0.1632  0.2544  0.2807  0.3011  0.3361  0.3024  0.2285  0.2910  0.1316   
102  0.2352  0.3208  0.4257  0.5201  0.4914  0.5950  0.7221  0.9039  0.9111   
130  0.3083  0.3794  0.5364  0.6173  0.7842  0.8392  0.9016  1.0000  0.8911   
36   0.1328  0.1303  0.0273  0.0644  0.0712  0.1204  0.0717  0.1224  0.2349   

         18      19      20      21      22      23      24      25      26  \
77   0.7988  0.8261  1.0000  0.9814  0.9620  0.9601  0.9118  0.9086  0.7931   
142  0.1151  0.3404  0.5562  0.6379  0.6553  0.7384  0.6534  0.5423  0.6877   
102  0.8723  0.7686  0.7326  0.5222  0.3097  0.3172  0.2270  0.1640  0.1746   
130  0.8753  0.7886  0.7156  0.7581  0.6372  0.3210  0.2076  0.2279  0.3309   
36   0.3684  0.3918  0.4925  0.8793  0.9606  0.8786  0.6905  0.6937  0.5674   

         27      28      29      30      31      32      33      34      35  \
77   0.5877  0.3474  0.4235  0.4633  0.3410  0.2849  0.2847  0.1742  0.0549   
142  0.7325  0.7726  0.8229  0.8787  0.9108  0.6705  0.6092  0.7505  0.4775   
102  0.1835  0.2048  0.1674  0.2767  0.3104  0.3399  0.4441  0.5046  0.2814   
130  0.2847  0.1949  0.1671  0.1025  0.1362  0.2212  0.1124  0.1677  0.1039   
36   0.6540  0.7802  0.7575  0.5836  0.6316  0.8108  0.9039  0.8647  0.6695   

         36      37      38      39      40      41      42      43      44  \
77   0.1192  0.1154  0.0855  0.1811  0.1264  0.0799  0.0378  0.1268  0.1125   
142  0.1666  0.3749  0.3776  0.2106  0.5886  0.5628  0.2577  0.5245  0.6149   
102  0.1681  0.2633  0.3198  0.1933  0.0934  0.0443  0.0780  0.0722  0.0405   
130  0.2562  0.2624  0.2236  0.1180  0.1103  0.2831  0.2385  0.0255  0.1967   
36   0.4027  0.2370  0.2685  0.3662  0.3267  0.2200  0.2996  0.2205  0.1163   

         45      46      47      48      49      50      51      52      53  \
77   0.0505  0.0949  0.0677  0.0259  0.0170  0.0033  0.0150  0.0111  0.0032   
142  0.5123  0.3385  0.1499  0.0546  0.0270  0.0380  0.0339  0.0149  0.0335   
102  0.0553  0.1081  0.1139  0.0767  0.0265  0.0215  0.0331  0.0111  0.0088   
130  0.1483  0.0434  0.0627  0.0513  0.0473  0.0248  0.0274  0.0205  0.0141   
36   0.0635  0.0465  0.0422  0.0174  0.0172  0.0134  0.0141  0.0191  0.0145   

         54      55      56      57      58      59 60  
77   0.0035  0.0169  0.0137  0.0015  0.0069  0.0051  R  
142  0.0376  0.0174  0.0132  0.0103  0.0364  0.0208  M  
102  0.0158  0.0122  0.0038  0.0101  0.0228  0.0124  M  
130  0.0185  0.0055  0.0045  0.0115  0.0152  0.0100  M  
36   0.0065  0.0129  0.0217  0.0087  0.0077  0.0122  R

In [118]:
df.shape

(208, 61)

In [119]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       208 non-null    float64
 1   1       208 non-null    float64
 2   2       208 non-null    float64
 3   3       208 non-null    float64
 4   4       208 non-null    float64
 5   5       208 non-null    float64
 6   6       208 non-null    float64
 7   7       208 non-null    float64
 8   8       208 non-null    float64
 9   9       208 non-null    float64
 10  10      208 non-null    float64
 11  11      208 non-null    float64
 12  12      208 non-null    float64
 13  13      208 non-null    float64
 14  14      208 non-null    float64
 15  15      208 non-null    float64
 16  16      208 non-null    float64
 17  17      208 non-null    float64
 18  18      208 non-null    float64
 19  19      208 non-null    float64
 20  20      208 non-null    float64
 21  21      208 non-null    float64
 22  22

In [120]:
# check for nan values
df.isna().sum().sort_values(ascending=False)

0     0
31    0
33    0
34    0
35    0
     ..
25    0
26    0
27    0
28    0
60    0
Length: 61, dtype: int64

In [121]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [122]:
df[60].value_counts() 

M    111
R     97
Name: 60, dtype: int64

111 samples metal, and 97 rocks 

In [123]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [124]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
168  0
187  0
157  0
91   1
200  0

In [125]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [126]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9       10      11      12      13      14      15      16      17  \
0  0.2111  0.1609  0.1582  0.2238  0.0645  0.0660  0.2273  0.3100  0.2999   
1  0.2872  0.4918  0.6552  0.6919  0.7797  0.7464  0.9444  1.0000  0.8874   
2  0.6194  0.6333  0.7060  0.5544  0.5320  0.6479  0.6931  0.6759  0.7551   
3  0.1264  0.0881  0.1992  0.0184  0.2261  0.1729  0.2131  0.0693  0.2281   
4  0.4459  0.4152  0.3952  0.4256  0.4135  0.4528  0.5326  0.7306  0.6193   

       18      19      20      21      22      23      24      25      26  \
0  0.5078  0.4797  0.5783  0.5071  0.4328  0.5550  0.6711  0.6415  0.7104   
1  0.8024  0.7818  0.5212  0.4052  0.3957  0.3914  0.3250  0.3200  0.3271   
2  0.8929  0.8619  0.7974  0.6737  0.4293  0.3648  0.5331  0.2413  0.5070   
3  0.4060  0.3973  0.2741  0.3690  0.5556  0.4846  0.3140  0.5334  0.5256   
4  0.2032  0.4636  0.4148  0.4292  0.5730  0.5399  0.3161  0.2285  0.6995   

       27      28      29      30      31      32      33      34      35  \
0  0.8080  0.6791  0.3857  0.1307  0.2604  0.5121  0.7547  0.8537  0.8507   
1  0.2767  0.4423  0.2028  0.3788  0.2947  0.1984  0.2341  0.1306  0.4182   
2  0.8533  0.6036  0.8514  0.8512  0.5045  0.1862  0.2709  0.4232  0.3043   
3  0.2520  0.2090  0.3559  0.6260  0.7340  0.6120  0.3497  0.3953  0.3012   
4  1.0000  0.7262  0.4724  0.5103  0.5459  0.2881  0.0981  0.1951  0.4181   

       36      37      38      39      40      41      42      43      44  \
0  0.6692  0.6097  0.4943  0.2744  0.0510  0.2834  0.2825  0.4256  0.2641   
1  0.3835  0.1057  0.1840  0.1970  0.1674  0.0583  0.1401  0.1628  0.0621   
2  0.6116  0.6756  0.5375  0.4719  0.4647  0.2587  0.2129  0.2222  0.2111   
3  0.5408  0.8814  0.9857  0.9167  0.6121  0.5006  0.3210  0.3202  0.4295   
4  0.4604  0.3217  0.2828  0.2430  0.1979  0.2444  0.1847  0.0841  0.0692   

       45      46      47      48      49      50      51      52      53  \
0  0.1386  0.1051  0.1343  0.0383  0.0324  0.0232  0.0027  0.0065  0.0159   
1  0.0203  0.0530  0.0742  0.0409  0.0061  0.0125  0.0084  0.0089  0.0048   
2  0.0176  0.1348  0.0744  0.0130  0.0106  0.0033  0.0232  0.0166  0.0095   
3  0.3654  0.2655  0.1576  0.0681  0.0294  0.0241  0.0121  0.0036  0.0150   
4  0.0528  0.0357  0.0085  0.0230  0.0046  0.0156  0.0031  0.0054  0.0105   

       54      55      56      57      58      59  
0  0.0072  0.0167  0.0180  0.0084  0.0090  0.0032  
1  0.0094  0.0191  0.0140  0.0049  0.0052  0.0044  
2  0.0180  0.0244  0.0316  0.0164  0.0095  0.0078  
3  0.0085  0.0073  0.0050  0.0044  0.0040  0.0117  
4  0.0110  0.0015  0.0072  0.0048  0.0107  0.0094

In [127]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [128]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9       10      11      12      13      14      15      16      17  \
67   0.0506  0.0906  0.2545  0.1464  0.1272  0.1223  0.1669  0.1424  0.1285   
14   0.2120  0.1640  0.1901  0.3026  0.2019  0.0592  0.2390  0.3657  0.3809   
164  0.2822  0.3691  0.3750  0.3927  0.3308  0.1085  0.1139  0.3446  0.5441   
179  0.2558  0.3078  0.3404  0.3400  0.3951  0.3352  0.2252  0.2086  0.2248   
19   0.5920  0.5832  0.5419  0.5472  0.5314  0.4981  0.6985  0.8292  0.7839   

         18      19      20      21      22      23      24      25      26  \
67   0.1857  0.1136  0.2069  0.0219  0.2400  0.2547  0.0240  0.1923  0.4753   
14   0.5929  0.6299  0.5801  0.4574  0.4449  0.3691  0.6446  0.8940  0.8978   
164  0.6470  0.7276  0.7894  0.8264  0.8697  0.7836  0.7140  0.5698  0.2908   
179  0.3382  0.4578  0.6474  0.6708  0.7007  0.7619  0.7745  0.6767  0.7373   
19   0.8215  0.9363  1.0000  0.9224  0.7839  0.5470  0.4562  0.5922  0.5448   

         27      28      29      30      31      32      33      34      35  \
67   0.7003  0.6825  0.6443  0.7063  0.5373  0.6601  0.8708  0.9518  0.9605   
14   0.4980  0.3333  0.2350  0.1553  0.3666  0.4340  0.3082  0.3024  0.4109   
164  0.4636  0.6409  0.7405  0.8069  0.8420  1.0000  0.9536  0.6755  0.3905   
179  0.7834  0.9619  1.0000  0.8086  0.5558  0.5409  0.4988  0.3108  0.2897   
19   0.3971  0.0882  0.2385  0.2005  0.0587  0.2544  0.2009  0.0329  0.1547   

         36      37      38      39      40      41      42      43      44  \
67   0.7712  0.6772  0.6431  0.6720  0.6035  0.5155  0.3802  0.2278  0.1522   
14   0.5501  0.4129  0.5499  0.5018  0.3132  0.2802  0.2351  0.2298  0.1155   
164  0.1249  0.3629  0.6356  0.8116  0.7664  0.5417  0.2614  0.1723  0.2814   
179  0.2244  0.0960  0.2287  0.3228  0.3454  0.3882  0.3240  0.0926  0.1173   
19   0.1212  0.2446  0.3171  0.3195  0.3051  0.0836  0.1266  0.1381  0.1136   

         45      46      47      48      49      50      51      52      53  \
67   0.0801  0.0804  0.0752  0.0566  0.0175  0.0058  0.0091  0.0160  0.0160   
14   0.0724  0.0621  0.0318  0.0450  0.0167  0.0078  0.0083  0.0057  0.0174   
164  0.2764  0.1985  0.1502  0.1219  0.0493  0.0027  0.0077  0.0026  0.0031   
179  0.0566  0.0766  0.0969  0.0588  0.0050  0.0118  0.0146  0.0040  0.0114   
19   0.0516  0.0073  0.0278  0.0372  0.0121  0.0153  0.0092  0.0035  0.0098   

         54      55      56      57      58      59  
67   0.0081  0.0070  0.0135  0.0067  0.0078  0.0068  
14   0.0188  0.0054  0.0114  0.0196  0.0147  0.0062  
164  0.0083  0.0020  0.0084  0.0108  0.0083  0.0033  
179  0.0032  0.0062  0.0101  0.0068  0.0053  0.0087  
19   0.0121  0.0006  0.0181  0.0094  0.0116  0.0063

# Using Deep Learning Model

## Model without Dropout Layer

In [129]:
import tensorflow as tf
from tensorflow import keras

In [130]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.6905 - accuracy: 0.5321
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6696 - accuracy: 0.6218
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6469 - accuracy: 0.6987
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6214 - accuracy: 0.7500
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5943 - accuracy: 0.7500
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5535 - accuracy: 0.7885
Epoch 7/100
20/20 [==============================] - 0s 6ms/step - loss: 0.5360 - accuracy: 0.7628
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4923 - accuracy: 0.8077
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4448 - accuracy: 0.8205
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4138 - accuracy: 0.8397
Epoch 11/

In [131]:
#model.evaluate(X_test, y_test)

Training Accuracy >>> Test Accuracy

In [132]:
#y_pred = model.predict(X_test).reshape(-1)
#print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
#y_pred = np.round(y_pred)
#print(y_pred[:10])

In [133]:
#y_test[:10]

In [134]:
#from sklearn.metrics import confusion_matrix , classification_report

#print(classification_report(y_test, y_pred))

### Model with Dropout Layer

In [135]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 3ms/step - loss: 0.6970 - accuracy: 0.4936
Epoch 2/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7079 - accuracy: 0.5449
Epoch 3/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7024 - accuracy: 0.4551
Epoch 4/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6986 - accuracy: 0.5064
Epoch 5/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6921 - accuracy: 0.5192
Epoch 6/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6865 - accuracy: 0.5321
Epoch 7/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6958 - accuracy: 0.5064
Epoch 8/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6793 - accuracy: 0.5962
Epoch 9/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6823 - accuracy: 0.5513
Epoch 10/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6875 - accuracy: 0.5705
Epoch 11/

In [136]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 4ms/step - loss: 0.4355 - accuracy: 0.7692


[0.43553924560546875, 0.7692307829856873]

Training Accuracy is still good but Test Accuracy Improved

In [137]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 3ms/step
[0.0024609  0.7915153  0.9144191  0.05611353 0.9996545  0.95766705
 0.3951249  0.9995616  0.05497704 0.99995416]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [138]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.73      0.89      0.80        27
           1       0.84      0.64      0.73        25

    accuracy                           0.77        52
   macro avg       0.78      0.76      0.76        52
weighted avg       0.78      0.77      0.77        52



**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**